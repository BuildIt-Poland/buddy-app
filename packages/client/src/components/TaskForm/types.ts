import { TaskInput, Task } from '@buddy-app/schema';

export interface Dictionary {
  BUTTON_TEXT: string;
  TITLE: {
    LABEL: string;
    REQUIRED: string;
  };
  DESCRIPTION: {
    LABEL: string;
    HELPER_TEXT: string;
    REQUIRED: string;
  };
}

export interface TaskFormProps {
  dictionary: Dictionary;
  loading: boolean;
  data?: Task;
  onSubmit: (input: TaskInput) => void;
}
