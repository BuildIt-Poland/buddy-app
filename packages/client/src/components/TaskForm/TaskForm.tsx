import React from 'react';
import { useForm } from 'react-hook-form';
import { TaskInput } from '@buddy-app/schema';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import CircularProgress from '@material-ui/core/CircularProgress';
import TextField from '@material-ui/core/TextField';
import RoundedButton from 'atoms/RoundedButton';
import { TaskFormProps } from './types';
import DICTIONARY from './dictionary';

const useStyles = makeStyles(theme => ({
  form: {
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
  },
  inputWrapper: {
    flex: 1,
  },
  addButton: {
    margin: theme.spacing(5, 0, 4),
  },
}));

const TaskForm: React.FC<TaskFormProps> = ({ loading, data, onSubmit }) => {
  const { addButton, inputWrapper, form } = useStyles();
  const { register, errors, handleSubmit } = useForm<TaskInput>();

  return (
    <form
      className={form}
      data-testid='form'
      noValidate
      onSubmit={handleSubmit(onSubmit)}>
      <Box className={inputWrapper}>
        <TextField
          inputProps={{ 'data-testid': 'title' }}
          inputRef={register({
            required: DICTIONARY.TITLE.REQUIRED,
          })}
          margin={'normal'}
          fullWidth
          label={DICTIONARY.TITLE.LABEL}
          name='title'
          autoFocus
          defaultValue={data && data.title}
          error={!!errors.title}
          helperText={errors.title && errors.title.message}
        />
        <TextField
          inputProps={{ 'data-testid': 'description' }}
          inputRef={register({
            required: DICTIONARY.DESCRIPTION.REQUIRED,
          })}
          margin={'normal'}
          fullWidth
          name='description'
          label={DICTIONARY.DESCRIPTION.LABEL}
          multiline
          autoComplete='current-password'
          defaultValue={data && data.description}
          error={!!errors.description}
          helperText={
            (errors.description && errors.description.message) ||
            DICTIONARY.DESCRIPTION.HELPER_TEXT
          }
        />
      </Box>
      <RoundedButton
        data-testid='add-task-button'
        type='submit'
        fullWidth
        variant='contained'
        color='primary'
        disabled={loading}
        className={addButton}>
        {loading ? (
          <CircularProgress
            data-testid='add-task-progress'
            variant={'indeterminate'}
            size={32}
          />
        ) : (
          <strong>{DICTIONARY.BUTTON_TEXT}</strong>
        )}
      </RoundedButton>
    </form>
  );
};

export default TaskForm;
