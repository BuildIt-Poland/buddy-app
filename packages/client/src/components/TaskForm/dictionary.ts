interface Dictionary {
  BUTTON_TEXT: string;
  TITLE: {
    LABEL: string;
    REQUIRED: string;
  };
  DESCRIPTION: {
    LABEL: string;
    HELPER_TEXT: string;
    REQUIRED: string;
  };
}

const DICTIONARY: Dictionary = {
  BUTTON_TEXT: 'Submit',
  TITLE: {
    LABEL: 'Title',
    REQUIRED: 'Title is required',
  },
  DESCRIPTION: {
    LABEL: 'Description',
    HELPER_TEXT: 'HTML syntax is accepted',
    REQUIRED: 'Description is required',
  },
};

export default DICTIONARY;
