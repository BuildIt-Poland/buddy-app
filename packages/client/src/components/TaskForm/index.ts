import TaskForm from './TaskForm';
export * from './types';

export default TaskForm;
