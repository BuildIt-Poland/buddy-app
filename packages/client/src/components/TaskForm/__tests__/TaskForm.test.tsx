import React from 'react';
import { create } from 'react-test-renderer';
import TaskForm from '../TaskForm';

jest.mock('@material-ui/core/Box', () => 'Box');
jest.mock('@material-ui/core/CircularProgress', () => 'CircularProgress');
jest.mock('@material-ui/core/TextField', () => 'TextField');
jest.mock('atoms/RoundedButton', () => 'RoundedButton');

describe('Component - TaskForm', () => {
  test('renders correctly', () => {
    const component = create(<TaskForm loading onSubmit={() => {}} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
