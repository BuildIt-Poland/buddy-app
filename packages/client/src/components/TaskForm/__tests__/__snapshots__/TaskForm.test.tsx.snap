// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - TaskForm renders correctly 1`] = `
<form
  className="makeStyles-form-1"
  data-testid="form"
  noValidate={true}
  onSubmit={[Function]}
>
  <Box
    className="makeStyles-inputWrapper-2"
  >
    <TextField
      autoFocus={true}
      error={false}
      fullWidth={true}
      inputProps={
        Object {
          "data-testid": "title",
        }
      }
      inputRef={[Function]}
      label="Title"
      margin="normal"
      name="title"
    />
    <TextField
      autoComplete="current-password"
      error={false}
      fullWidth={true}
      helperText="HTML syntax is accepted"
      inputProps={
        Object {
          "data-testid": "description",
        }
      }
      inputRef={[Function]}
      label="Description"
      margin="normal"
      multiline={true}
      name="description"
    />
  </Box>
  <RoundedButton
    className="makeStyles-addButton-3"
    color="primary"
    data-testid="add-task-button"
    disabled={true}
    fullWidth={true}
    type="submit"
    variant="contained"
  >
    <CircularProgress
      data-testid="add-task-progress"
      size={32}
      variant="indeterminate"
    />
  </RoundedButton>
</form>
`;
